### PR TITLE
Bring Renderer Into Conformity With Changes Made on NDP

### PR DESCRIPTION
**Pull request by hummusonrails**
This PR brings in the changes that were previously made in NDP in the `MarkdownPipeline` into the gem. It includes: `services`, `filters`, `models` and new spec testing.

### CODE REVIEW

**Comment by hummusonrails** [comment]
@fabianrbz, The changes that I implemented in this PR currently fix all the outstanding test errors, but it still does not address the `uninitialized constant` error in a Rails prod app. The methods all work as expected when used independently outside of a Rails app. Would appreciate some insight here. Thanks.

**Comment by hummusonrails** [comment]
@fabianrbz, I've tested these changes both in the testing suite, in NDP and by itself, and they all seem to work.

File: lib/nexmo_markdown_renderer/filters/code_filter.rb [hunk 1]
```diff
@@ -1,56 +1,60 @@
...
-        config = config['config'].split('.').inject(configs) { |h, k| h[k] }
+module Nexmo
+  module Markdown
+    class CodeFilter < Banzai::Filter
+      def call(input)
+        input.gsub(/(?!.*snippet)```code(.+?)```/m) do |_s|
+          config = YAML.safe_load($1)
+    
+          if config['config']
+            configs = YAML.load_file("#{Rails.root}/config/code_examples.yml")
```
Comment (fabianrbz) [inline lib/nexmo_markdown_renderer/filters/code_filter.rb]: I don't think this block is needed, `#{Rails.root}/config/code_examples.yml`  does not even exist in `nexmo-developer`

File: lib/nexmo_markdown_renderer/filters/code_filter.rb [hunk 2]
```diff
@@ -1,56 +1,60 @@
...
+            config = config['config'].split('.').inject(configs) { |h, k| h[k] }
+          end
+    
+          code = File.read("#{Rails.root}/#{config['source']}")
```
Comment (fabianrbz) [inline lib/nexmo_markdown_renderer/filters/code_filter.rb]: I think we should not rely on the gem being part of a Rails application, instead, we should use an `ENV` variable that points to the docs and use that as the `root_path`

File: lib/nexmo_markdown_renderer/filters/code_filter.rb [hunk 3]
```diff
@@ -1,56 +1,60 @@
...
+      end
+    
+      def language_to_lexer(language)
+        language = language_to_lexer_name(language)
+        return Rouge::Lexers::PHP.new({ start_inline: true }) if language == 'php'
+        Rouge::Lexer.find(language.downcase) || Rouge::Lexer.find('text')
+      end
+    
+      def language_configuration
+        @language_configuration ||= YAML.load_file("#{Rails.root}/config/code_languages.yml")
```
Comment (fabianrbz) [inline lib/nexmo_markdown_renderer/filters/code_filter.rb]: shouldn't we move that file over to this gem?

File: lib/nexmo_markdown_renderer/filters/code_snippet_list_filter.rb [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
+    
+          # Top level blocks come before subfolder
+          @blocks.sort_by! do |b|
+            sort_array = []
+            sort_array << (b.category ? 1 : 0)
+            sort_array << b.navigation_weight
+            sort_array
+          end
+    
+          erb = File.read("#{Rails.root}/app/views/code_snippets/list/plain.html.erb")
```
Comment (fabianrbz) [inline lib/nexmo_markdown_renderer/filters/code_snippet_list_filter.rb]: we should move this view over

File: lib/nexmo_markdown_renderer/filters/code_filter.rb (hunk 1 above)
Comment (hummusonrails) [reply lib/nexmo_markdown_renderer/filters/code_filter.rb]: 👍

File: lib/nexmo_markdown_renderer/filters/code_filter.rb (hunk 2 above)
Comment (hummusonrails) [reply lib/nexmo_markdown_renderer/filters/code_filter.rb]: 👍

File: lib/nexmo_markdown_renderer/filters/code_filter.rb (hunk 3 above)
Comment (hummusonrails) [reply lib/nexmo_markdown_renderer/filters/code_filter.rb]: 👍

File: lib/nexmo_markdown_renderer/filters/code_snippet_list_filter.rb (hunk 4 above)
Comment (hummusonrails) [reply lib/nexmo_markdown_renderer/filters/code_snippet_list_filter.rb]: 👍

File: spec/filters/anchor_filter_spec.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
```
Comment (fabianrbz) [inline spec/filters/anchor_filter_spec.rb]: we should drop the `rails_helper` and use `spec_helper` instead.

File: spec/filters/code_snippets_filter_spec.rb [hunk 6]
```diff
@@ -13,11 +17,11 @@
 
   it 'creates correct html output with correct input' do
     expect(SecureRandom).to receive(:hex).at_least(:once).and_return('ID123456')
-    allow_any_instance_of(CodeSnippetsFilter).to receive(:render_single_snippet).and_return('')
+    allow_any_instance_of(Nexmo::Markdown::CodeSnippetsFilter).to receive(:render_single_snippet).and_return('')
```
Comment (fabianrbz) [inline spec/filters/code_snippets_filter_spec.rb]: you can use `described_class` instead :)

File: spec/filters/code_snippet_filter_spec.rb [hunk 7]
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe CodeSnippetFilter do
+RSpec.describe Nexmo::Markdown::CodeSnippetFilter do
```
Comment (fabianrbz) [inline spec/filters/code_snippet_filter_spec.rb:3]: some of these specs are using snapshots, can you add the path to the snapshots to the `spe_helper.rb` and copy over the snapshots that were generated in NDP?

**Review by fabianrbz — CHANGES_REQUESTED** [review]
@benhayehudi this looks good. I left some minor comments.

**Comment by fabianrbz** [comment]
Note: the gem as it is now will only work within NDP and not as standalone

File: lib/nexmo_markdown_renderer/services/code_snippet_renderer/android.rb [hunk 8]
```diff
@@ -11,11 +11,11 @@ def self.run_command(_command, _filename, _file_path)
     end
```
Comment (fabianrbz) [inline lib/nexmo_markdown_renderer/services/code_snippet_renderer/android.rb]: all these services should be namespaced

File: spec/filters/anchor_filter_spec.rb (hunk 5 above)
Comment (hummusonrails) [reply spec/filters/anchor_filter_spec.rb]: 👍

File: spec/filters/code_snippets_filter_spec.rb (hunk 6 above)
Comment (hummusonrails) [reply spec/filters/code_snippets_filter_spec.rb]: 👍

File: spec/filters/code_snippet_filter_spec.rb (hunk 7 above)
Comment (hummusonrails) [reply spec/filters/code_snippet_filter_spec.rb:3]: Just verified that the snapshots in the gem are the exact same as the ones currently in NDP.

File: lib/nexmo_markdown_renderer/services/code_snippet_renderer/android.rb (hunk 8 above)
Comment (hummusonrails) [reply lib/nexmo_markdown_renderer/services/code_snippet_renderer/android.rb]: 👍

**Review by fabianrbz — APPROVED** [review]
:shipit:

**Pull request merged**